[DR-3042] Share common HTTP clients

okotsopoulos opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-3042

While investigating recurring memory leaks in TDR, I found that we are not following best practices when constructing HTTP clients.  Visualization of a recent memory leak in a TDR prod API pod:

<img width="1196" alt="Screenshot 2023-05-18 at 12 07 24 PM" src="https://github.com/DataBiosphere/jade-data-repo/assets/79769153/277334fd-5bf8-46da-a895-284c33e7cbbb">

Timely discussion in Slack: https://broadinstitute.slack.com/archives/C03T815L316/p1684422897621399

A few instances of DSP projects which presently share clients safely:
- https://github.com/DataBiosphere/terra-workspace-manager/blob/3970810228cb235d507dcdb4a3c54f33fff5c463/service/src/main/java/bio/terra/workspace/service/iam/SamService.java#L106
- https://github.com/DataBiosphere/terra-data-catalog/blob/main/common/src/main/java/bio/terra/catalog/rawls/RawlsClient.java 

My developer environment is up to date with these latest changes:
- UI https://jade-ok.datarepo-dev.broadinstitute.org/
- Overview of my API deployment https://console.cloud.google.com/kubernetes/deployment/us-central1/dev-master/ok/ok-jade-datarepo-api/overview?authuser=1&project=broad-jade-dev
- 
  It's certainly lower traffic than other environments and my changes haven't been on there long, so it's hard to draw sweeping conclusions.  But I do see that the TDR dev API pods show signs of memory leaks, so if we were to merge this in we'd likely get an early indication of its impact before it made it to prod:

<img width="1477" alt="Screenshot 2023-05-18 at 5 46 13 PM" src="https://github.com/DataBiosphere/jade-data-repo/assets/79769153/a951b0c8-9d0c-4a6a-bbb5-085e8fc23ed8">

**Sam**

[OkHttpClient docs](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-ok-http-client/) used for Sam calls:
> OkHttp performs best when you create a single OkHttpClient instance and reuse it for all of your HTTP calls. This is because each client holds its own connection pool and thread pools. Reusing connections and threads reduces latency and saves memory. Conversely, creating a client for each request wastes resources on idle pools.

We did not follow those recommendations and implicitly constructed new `OkHttpClient`s on every call.

Now, Sam APIs share a common `OkHttpClient`.  Tokens are stored on the wrapper `ApiClient` and not the shared `Client`.  Sam APIs may build new `OkHttpClient`s to handle custom configuration, but they will all share common connection pools and thread pools.

**RBS**

[Oracle Client docs](https://docs.oracle.com/javaee/7/api/javax/ws/rs/client/Client.html) used for RBS calls:
> Clients are heavy-weight objects that manage the client-side communication infrastructure. Initialization as well as disposal of a Client instance may be a rather expensive operation. It is therefore advised to construct only a small number of Client instances in the application. Client instances must be properly closed before being disposed to avoid leaking resources.

We did not follow those recommendations and implicitly constructed new Clients on every call without closing them.

Now, RBS APIs share a common `Client`, though we don't bother to close it as it will be closed on app shut-down anyway.  Tokens are stored on the wrapper `ApiClient` and not the shared `Client`.

**Others**

I also found several instances where we were not using the shared `RestTemplate` bean, but could have been.  I flagged swaps that I wasn't confident in for reviewers.

**Follow-On Work**

I may not get to these in time for the next release train, so I think if we're comfortable with these current changes we shouldn't hold the PR and instead pursue a follow-on for the rest and any further improvements.

- [x] `PolicyApiService` should use a shared `Client`
- [x] `Oauth2ApiController` can use `RestTemplate` bean
- [ ] `OpenIDConnectConfiguration` can use `RestTemplate` bean (actually will hold off, looking at this I think it might need its own?)

okotsopoulos committed:
[DR-3042] Sam ApiClients share a common OkHttpClient

Previously, we would create a new one for each request, which goes against OkHttpClient recommendations.

okotsopoulos committed:
BufferService APIs share a common Client

Drive-by: removed unused method.

okotsopoulos committed:
I think that RawlsClient can use common RestTemplate bean

okotsopoulos committed:
I think that BardClient can use common RestTemplate bean

Drive-by: looked like we were passing duplicate acceptable media types to our headers.
Refactored BardClientTest to use JUnit 5 and remove need for BardClient spy.

okotsopoulos commented on src/main/java/bio/terra/app/usermetrics/BardClient.java:
```diff
@@ -41,13 +40,12 @@ public class BardClient {
 
   @Autowired
-  public BardClient(UserMetricsConfiguration metricsConfig) {
-    this.restTemplate = new RestTemplate();
-    restTemplate.setRequestFactory(new HttpComponentsClientHttpRequestFactory());
```
Flagging for reviewers: is there a reason that we weren't using the common `RestTemplate` bean before?

nmalfroy replied:
Nope! I think this might have predated having the shared bean.  Good catch!

okotsopoulos commented on src/main/java/bio/terra/service/rawls/RawlsClient.java:
```diff
@@ -27,10 +26,9 @@ public class RawlsClient {
-  public RawlsClient(RawlsConfiguration rawlsConfiguration) {
+  public RawlsClient(RawlsConfiguration rawlsConfiguration, RestTemplate restTemplate) {
     this.rawlsConfiguration = rawlsConfiguration;
-    this.restTemplate = new RestTemplate();
-    restTemplate.setRequestFactory(new HttpComponentsClientHttpRequestFactory());
```
Flagging for reviewers: is there a reason that we weren't using the common `RestTemplate` bean before?

nmalfroy replied:
Nope, I think that got copy pasted from the Bard client

okotsopoulos replied:
🍝

okotsopoulos commented on src/main/java/bio/terra/app/usermetrics/BardClient.java:
```diff
@@ -41,13 +40,12 @@ public class BardClient {
 
     this.headers = new HttpHeaders();
     headers.setContentType(MediaType.APPLICATION_JSON);
-    headers.setAccept(Arrays.asList(MediaType.APPLICATION_JSON, MediaType.APPLICATION_JSON));
```
I assumed that these duplicate headers were an oversight.

nmalfroy replied:
Yeah, that seems like a bug

okotsopoulos commented on src/main/java/bio/terra/service/resourcemanagement/BufferService.java:
```diff
@@ -47,51 +49,23 @@ public BufferService(
-   *
-   * @return PoolInfo
-   */
-  public PoolInfo getPoolInfo() {
```
Unused code.

nmalfroy reviewed (approved):
Thanks for this, it looks great! I feel like we might have a similar thing latently in the Azure SDK code but that's a whole different can of worms and is worth investigating separately

okotsopoulos commented on src/main/java/bio/terra/service/resourcemanagement/BufferService.java:
```diff
@@ -105,7 +79,7 @@ public ResourceInfo handoutResource(boolean enableSecureMonitoring) {
     logger.info("Using request ID: {} to get project from RBS", handoutRequestId);
     HandoutRequestBody requestBody = new HandoutRequestBody().handoutRequestId(handoutRequestId);
     try {
-      BufferApi bufferApi = bufferApi(bufferServiceConfiguration.getInstanceUrl());
```
This seemed redundant -- we can access the config when we create the clients and only set the base path in one place.

okotsopoulos committed:
TPS APIs share a common Client

Drive-by: consolidate API creation so that we're only setting the base path in one place.

samanehsan reviewed (approved):
Thanks for fixing this (and including drive-by improvements)!

okotsopoulos committed:
Oauth2ApiController can use RestTemplate bean

Refactored Oauth2ApiControllerTest to use JUnit 5 with context configuration (test slicing) to opt in the only controller we're testing.
We then must mock Oauth2ApiController's beans but the upside is faster test start-up.

okotsopoulos committed:
Drive-by: deduplicate acceptable headers in SamFixtures, DataRepoClient

okotsopoulos commented on src/test/java/bio/terra/app/controller/Oauth2ApiControllerTest.java:
Hmmm… I thought this refactoring to use `ContextConfiguration` would reduce test suite run time, but it actually doubled from 10 -> 20 seconds.

Any ideas?

nmalfroy replied:
Is that where the slowdown comes from?  Could it be the mock initialization happening before each test rather than before the class?  (I kinda doubt it but figured I'd mention)

nmalfroy replied:
Actually, does the slowdown come from the `ContextConfiguration` reuse or from upgrading junit to 5?

pshapiro4broad commented on src/test/java/bio/terra/service/auth/iam/sam/SamApiServiceTest.java:
```diff
@@ -2,63 +2,81 @@
+    // we could have many HttpClients built to support different configuration that all reuse the
+    // same connection pool.
+    // As a proxy check, we make sure that the same connection pool is reused.
+    List<ConnectionPool> connectionPools =
+        apiClients.stream().map(ac -> ac.getHttpClient().connectionPool()).distinct().toList();
+    assertThat("All Sam ApiClients share a connection pool", connectionPools, hasSize(1));
```
minor suggestion, you could use `count()` here instead of `toList() ... hasSize(1)`
```suggestion
    int numConnectionPools =
        apiClients.stream().map(ac -> ac.getHttpClient().connectionPool()).distinct().count();
    assertThat("All Sam ApiClients share a connection pool", numConnectionPools, is(1));
```
